### ISSUE TITLE
mouse.fix.helper does not uninstall when removing the application [bug]

### ISSUE THREAD

**Issue by ShayanYaseen** [noah-nuebling/mac-mouse-fix#128]
## Description

Hi 

I installed this application a few months back and then uninstalled it. However when I ran launchctl list today I found mouse.fix.helper in the list (with no pid and status 78). How can I remove this service from my mac ? I have already tried reinstalling this application and then removing it again.

## Steps to reproduce

1. Install the application
2. Uninstall it
3. Reboot your mac or not and then run launchctl list
4. Find mouse.fix.helper in the launchd list results

## Context

Mac Mouse Fix Version: 2.0.0 Beta 9
macOS Version: 11.4
Mouse Model: –

## Additional Comments

–

---

### Console Logs

–

### Crash Reports

–

**Comment by ShayanYaseen** [comment]
I removed the `com.nuebling.mac-mouse-fix.helper.plist` in /Library/LaunchAgents from my system which should prevent the service from launching but I still don't know why it did'nt get uninstalled with the rest of the app.

Anyway I hope you might look into this isssue and fix it on your end if my case is not some spooky anomaly.

Thanks

P.S Could you provide me with a list of locations where your application creates files and services so that I might look into them for remanants.

**Comment by noah-nuebling** [comment]
Hey @ShayanYaseen, thanks for reaching out!

I recommend [AppCleaner by freemacsoft](https://freemacsoft.net/appcleaner/) for removing leftovers from the library. I tried to make Mac Mouse Fix leave behind no residue but it’s not feasible with how macOS works unfortunately. Also, most Mac Mouse Fix related files in your library weren’t created by Mac Mouse Fix directly, but by the system on behalf of it. I don’t know where all of them are, either, unfortunately, so your best bet is a cleaner app like AppCleaner by freemacsoft or CleanMyMacX.

**The details**

The launchd.plist file at ~/Library/LaunchAgents/com.nuebling.mac-mouse-fix.helper.plist should be deleted by Mac Mouse Fix when you disable it or when you delete the app while it’s enabled. 
This definitely used to work when I built that feature but that has been a while ago so maybe it broke in the mean time. I’ll try to look into it at some point soon. But it’s not a big priority right now because there will be tons of other files left in your system anyways as explained below. Also it should have absolutely no noticeable effect if the launchd.plist is still there after you delete the app, except that a `mouse.fix.helper` entry it will show up when you run the `launchctl List` terminal command.

Another possible explanation for you issue I can think of off the top of my head aside from that my code is broken is that your ~/Library/LaunchAgents folder maybe has broken permissions.  A few users had trouble enabling Mac Mouse Fix recently and for a few of them it was because their LaunchAgents folder had permissions set to “read only”. No idea why. But that could also explain why the launchd.plist file wasn’t deleted from your LaunchAgent folder when you deleted Mac Mouse Fix.

As for the other files that Mac Mouse Fix creates. I really tried to leave as little junk behind in the system as possible, but macOS makes this pretty much impossible unfortunately.

Here’s what’s going on:

Mac Mouse Fix only creates two files directly:
- ~/Library/Application Support/com.nuebling.mac-mouse-fix/config.plist
  - For holding your preferences and some other stuff. I tried to put this inside the app bundle but changing the contents of a bundle after it’s signed totally breaks the app signing, which will make the macOS security mechanisms go haywire. So I had to put it into the library.
- ~/LaunchAgents/com.nuebling.mouse-fix-helper.plist
  - This was absolutely necessary for starting Mac Mouse Fix Helper at launch in a robust way when MMF was still a PrefPane. Now that it’s an app I could use a different mechanism which won’t create a file (SMLoginItemSetEnabled). But it would leave traces in a system database instead as far as I understand, and it has some other drawbacks. Overall it’s just not worth transitioning to it now.
 
When I moved the config file outside of the App Bundle, I built a mechanism where Mac Mouse Fix Helper can detect when the main app is deleted. It will then delete both files it created from your library. This of course only works when the helper is running (when MMF is enabled).

Aside from that there is no (feasible) way for Mac Mouse Fix to automatically remove the junk it left behind in the library when you uninstall it. macOS simply isn’t made for that.

And I’ve noticed another problem in the meantime: macOS creates tons of extra files related to MMF in the library that I have zero control over or knowledge about. Things like caches, saved application state, crash logs, etc.

I simply don’t know where these files are, because Mac Mouse Fix didn’t even create them itself.

Maybe I’ll get around to finding all the other possible locations for junk files and also uninstalling them automatically. But I feel like it’s not worth it at that point.

To make this work really well, and not only when you delete Mac Mouse Fix while it is enabled, I’d have to create a separate app bundled in MMF, which is just there for uninstalling. I’d then have to maintain a separate launchd.plist file which tells the system “If there are any files added / changed / removed in the folder where MMF is installed then start this separate MMF uninstaller app”. And then I’d have to somehow update this launchd.plist file when the installation folder of MMF changes and so on.

At that point I’d basically be reimplementing an App like AppCleaner, but only for MMF which feels not worth it. Also there’d be a performance cost whenever you edit or move any file from / to the MMF install folder, because your system would spin up the MMF uninstaller app every time. That is something people probably don’t want either. And when you think this through further, if all apps did this, then that would be feasible at all, as hundreds of uninstaller apps would spin up every time that you update an app in your applications folder or add / remove a new app to the applications folder. At that point your system might slow down noticeably.

Long story short, macOS just sucks in this respect because it forces devs to put stuff in the library, then it puts even more extra stuff in the library on behalf of the app which the apps’ devs don’t even know about, and then it gives devs and users no good way to remove this stuff when the user uninstalls an app.

But luckily there is a nice solution after all and that’s using an app like [AppCleaner by freemacsoft](https://freemacsoft.net/appcleaner/) or CleanMyMacX.

It will detect when you move an app to the trash and then find left behind files in the library. Then it’ll give you the option to trash them along side the app you just trashed at the press of a button.

AppCleaner is also free and super easy to use. I really like it, and I think you will, too.

Hope that helps, and let me know if you have any more questions.
